pkg/cmd/roachtest/tests/tpcc.go

```diff
@@ -267,21 +286,32 @@ func runTPCC(ctx context.Context, t test.Test, c cluster.Cluster, opts tpccOptio
 			if len(workloadInstances) > 1 {
 				statsPrefix = fmt.Sprintf("workload_%d.", i)
 			}
-			t.WorkerStatus(fmt.Sprintf("running tpcc worker=%d warehouses=%d ramp=%s duration=%s on %s (<%s)",
-				i, opts.Warehouses, rampDur, opts.Duration, pgURLs[i], time.Minute))
-			cmd := fmt.Sprintf(
-				"./cockroach workload run tpcc --warehouses=%d --histograms="+t.PerfArtifactsDir()+"/%sstats.json "+
-					opts.ExtraRunArgs+" --ramp=%s --duration=%s --prometheus-port=%d --pprofport=%d %s %s",
-				opts.Warehouses,
-				statsPrefix,
-				rampDur,
-				opts.Duration,
-				workloadInstances[i].prometheusPort,
-				workloadPProfStartPort+i,
-				workloadInstances[i].extraRunArgs,
-				pgURLs[i],
-			)
-			return c.RunE(ctx, option.WithNodes(workloadNode), cmd)
+			l.Printf("running tpcc worker=%d warehouses=%d ramp=%s duration=%s on %s (<%s)",
+				i, opts.Warehouses, rampDur, opts.Duration, pgURLs[i], time.Minute)
+
+			histogramsPath := fmt.Sprintf("%s/%sstats.json", t.PerfArtifactsDir(), statsPrefix)
+			cmd := roachtestutil.NewCommand("%s workload run tpcc", test.DefaultCockroachPath).
+				MaybeFlag(opts.DB != "", "db", opts.DB).
+				Flag("warehouses", opts.Warehouses).
+				Flag("histograms", histogramsPath).
+				Flag("ramp", rampDur).
+				Flag("duration", opts.Duration).
+				Flag("prometheus-port", workloadInstances[i].prometheusPort).
+				Flag("pprofport", workloadPProfStartPort+i).
+				Arg(opts.ExtraRunArgs).
+				Arg(workloadInstances[i].extraRunArgs).
+				Arg(pgURLs[i])
+
+			err := c.RunE(ctx, option.WithNodes(workloadNode), cmd.String())
+			// Don't fail the test if we are running the workload throughout
+			// the entire test. Canceling the context just means that the
+			// test already failed, or that the test finished (and therefore
+			// this function should not return an error).
+			if ctx.Err() != nil && opts.Duration == 0 {
+				return nil
```

Comment:
I couldn't immediately find examples of tests running the workload indefinitely (i.e., `Duration == 0`).

Reply:
Yep, the background tpcc workload introduced in this PR should be the only one (hence this logic needed to be added).